pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java

```diff
@@ -979,7 +981,32 @@ public void testLookupThrottlingForClientByClient() throws Exception {
         EventLoopGroup eventLoop = EventLoopUtil.newEventLoopGroup(20, false,
                 new DefaultThreadFactory("test-pool", Thread.currentThread().isDaemon()));
         long reqId = 0xdeadbeef;
-        try (ConnectionPool pool = new ConnectionPool(conf, eventLoop)) {
+
+        Semaphore clientCnxSemaphore = new Semaphore(1);
+        AtomicInteger count = new AtomicInteger(2);
+        try (ConnectionPool pool = new ConnectionPool(conf, eventLoop, () -> new ClientCnx(conf, eventLoop) {
+            @Override
+            protected void handleLookupResponse(CommandLookupTopicResponse lookupResult) {
+                try {
+                    clientCnxSemaphore.acquire();
```

Comment:
If the client handled the first request quickly enough, there would never be three concurrent requests.

I think we just need to block the client so that the client doesn't process the response quickly. Maybe CountDownLatch is a better fit, just lick this:

```java
// request 1 :
countDownLatch.await()
// request 2 :
countDownLatch.await()
// request 3 :
countDownLatch.await()
// After countDown is executed, we have built three concurrent requests
countDownLacth.countDown()
```

Reply:
But if so, I think there still exists a small chance that  all of the three requests can request success, which will make the test fail. 

Because they are request concurrently,  then one of the requests, e.g., request3, still have the chance  that begin sending request after request1 and request2 completed @poorbarcode

Reply:
> Because they are request concurrently, then one of the requests, e.g., request3, still have the chance that begin sending request after request1 and request2 completed

No, Before we execute `countDownLacth.countDown`, the third request has already failed. see:

https://github.com/apache/pulsar/blob/be57e714c4b9de4e34f159285a44bd935b1d5367/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java#L790

I think you misunderstood me，Here's what I'm trying to say:

```
// override the clientCnx
static class CountDownLatchCnx extends ClientCnx{
    @Override
    void handleLookupResponse(){
        // countDownLatch.await();
    }
}

consumer1.lookup(...)
consumer2.lookup(...)
consumer3.lookup(...) // this request will be fail.

countDownLatch.countDown();
```

Reply:
Oh, I get your point now @poorbarcode , thanks for your explantion. 

I have thought using `CountDownLatch` here before,   But `CountDownLatch` has a problem that it can not reset it's count, which means we can not reuse it with multi cases in this test.  So I chosen using `Semaphore` here.

Reply:
In current implementation: `semaphore` will always permit the first request, there would never be three concurrent requests. Maybe we can use `await` & `notifyAll` instead `CountDownLatch`

Reply:
`await` & `notifyAll` Still have the same problem that is thread can only `await` once. 

> CountDownLatch has a problem that it can not reset it's count, which means we can not reuse it with multi cases in this test

Maybe  we can used an array `CountDownLatch[]` to reset a new CountDownLatch object.  I have updated it, please take a look again @poorbarcode, thanks.